File: client-random.cpp
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <string>
#include <cstdlib>
#include <ctime>

#pragma comment(lib, "ws2_32.lib")

using namespace std;

string generateRandomMessage(int length) {
    const string chars = "abcdefghijklmnopqrstuvwxyz ";
    string message;
    for (int i = 0; i < length; i++) {
        message += chars[rand() % chars.length()];
    }
    return message;
}

int main(int argc, char* argv[]) {
    WSADATA wsaData;
    int result;
    
    // Seed random number generator
    srand(time(NULL));
    
    // Get server name from command line or use localhost
    string serverName = "127.0.0.1";
    if (argc > 1) {
        serverName = argv[1];
    }
    
    int serverPort = 12000;
    
    // Initialize Winsock
    result = WSAStartup(MAKEWORD(2, 2), &wsaData);
    if (result != 0) {
        cerr << "WSAStartup failed: " << result << endl;
        return 1;
    }
    
    // Generate random number of messages (1-10)
    int numMessages = (rand() % 10) + 1;
    cout << "Client will send " << numMessages << " random messages to " 
         << serverName << ":" << serverPort << endl;
    
    for (int i = 0; i < numMessages; i++) {
        // Create socket
        SOCKET clientSocket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
        if (clientSocket == INVALID_SOCKET) {
            cerr << "Socket creation failed: " << WSAGetLastError() << endl;
            WSACleanup();
            return 1;
        }
        
        // Setup server address
        sockaddr_in serverAddr;
        serverAddr.sin_family = AF_INET;
        serverAddr.sin_port = htons(serverPort);
        
        // Convert IP address
        if (inet_pton(AF_INET, serverName.c_str(), &serverAddr.sin_addr) <= 0) {
            cerr << "Invalid address: " << serverName << endl;
            closesocket(clientSocket);
            WSACleanup();
            return 1;
        }
        
        // Connect to server
        if (connect(clientSocket, (sockaddr*)&serverAddr, sizeof(serverAddr)) == SOCKET_ERROR) {
            cerr << "Connection failed: " << WSAGetLastError() << endl;
            cerr << "Make sure the server is running at " << serverName << ":" << serverPort << endl;
            closesocket(clientSocket);
            WSACleanup();
            return 1;
        }
        
        // Generate and send random message
        int messageLength = (rand() % 21) + 10; // 10-30 characters
        string message = generateRandomMessage(messageLength);
        
        cout << "\n[Message " << (i+1) << "/" << numMessages << "] Sending: " << message << endl;
        send(clientSocket, message.c_str(), message.length(), 0);
        
        // Receive response
        char buffer[1024] = {0};
        int bytesReceived = recv(clientSocket, buffer, sizeof(buffer), 0);
        
        if (bytesReceived > 0) {
            string response(buffer, bytesReceived);
            cout << "[Message " << (i+1) << "/" << numMessages << "] From Server: " << response << endl;
        }
        
        closesocket(clientSocket);
    }
    
    cout << "\n✓ Successfully sent all " << numMessages << " messages. Client terminating." << endl;
    
    WSACleanup();
    return 0;
}

File: client-random.js
const net = require('net');

// Get server name from command line or use localhost
const serverName = process.argv[2] || 'localhost';
const serverPort = 12000;

function generateRandomMessage(length) {
    const chars = 'abcdefghijklmnopqrstuvwxyz ';
    let message = '';
    for (let i = 0; i < length; i++) {
        message += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return message;
}

// Generate random number of messages (1-10)
const numMessages = Math.floor(Math.random() * 10) + 1;
console.log(`Client will send ${numMessages} random messages to ${serverName}:${serverPort}`);

let messagesSent = 0;

function sendMessage(messageNum) {
    const client = new net.Socket();

    client.connect(serverPort, serverName, () => {
        // Generate and send random message
        const messageLength = Math.floor(Math.random() * 21) + 10; // 10-30 characters
        const message = generateRandomMessage(messageLength);

        console.log(`\n[Message ${messageNum}/${numMessages}] Sending: ${message}`);
        client.write(message);
    });

    client.on('data', (data) => {
        console.log(`[Message ${messageNum}/${numMessages}] From Server: ${data.toString()}`);
        client.destroy();

        messagesSent++;

        // Send next message or finish
        if (messagesSent < numMessages) {
            sendMessage(messagesSent + 1);
        } else {
            console.log(`\n✓ Successfully sent all ${numMessages} messages. Client terminating.`);
        }
    });

    client.on('error', (err) => {
        console.error('ERROR:', err.message);
        console.error(`Make sure the server is running at ${serverName}:${serverPort}`);
        process.exit(1);
    });
}

// Start sending messages
sendMessage(1);
